hq_superset/hq_domain.py

```diff
@@ -64,6 +77,28 @@ def ensure_domain_selected():
         return redirect(url_for('SelectDomainView.list', next=request.url))
 
 
+def sync_user_domain_role():
+    if is_user_admin() or (
+        request.url_rule
+        and request.url_rule.endpoint in DOMAIN_EXCLUDED_VIEWS
+    ):
+        return
+    if _domain_role_expired():
+        _sync_domain_role()
+
+
+def _domain_role_expired():
+    if not session.get(SESSION_DOMAIN_ROLE_LAST_SYNCED_AT):
+        return True
+
+    time_since_last_sync = datetime_utcnow_naive() - session[SESSION_DOMAIN_ROLE_LAST_SYNCED_AT]
+    return time_since_last_sync >= timedelta(minutes=USER_DOMAIN_ROLE_EXPIRY)
+
+
+def _sync_domain_role():
+    DomainSyncUtil(superset.appbuilder.sm).sync_domain_role(g.hq_domain)
```

Comment:
@mkangia 
What happens if the domain role cannot sync during this routine? Is it fine to keep the user logged in in the current state?

Reply:
This turned out to be a little challenging, but I got it working like this
https://github.com/dimagi/commcare-analytics/pull/72/commits/a1959294970440ed801501d651b7e9a64da8d255

![Screenshot from 2024-11-05 00-58-37](https://github.com/user-attachments/assets/b229d609-97e0-4dea-926d-7a182be99748)